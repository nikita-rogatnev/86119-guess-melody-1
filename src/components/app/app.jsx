import React, {Component} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";

import {ActionCreator} from "../../reducer";
import ArtistQuestionScreen from "../artist-question-screen/artist-question-screen";
import QuestionGenreScreen from "../genre-question-screen/genre-question-screen";
import WelcomeScreen from "../welcome-screen/welcome-screen";

const Type = {
  ARTIST: `game--artist`,
  GENRE: `game--genre`,
};

class App extends Component {
  _getScreen(question) {
    if (!question) {
      const {
        maxMistakes,
        gameTime,
        onWelcomeScreenClick,
      } = this.props;

      return <WelcomeScreen
        errorCount={maxMistakes}
        gameTime={gameTime}
        onClick={onWelcomeScreenClick}
      />;
    }

    const {
      onUserAnswer,
      mistakes,
      maxMistakes,
    } = this.props;

    switch (question.type) {
      case `genre`:
        return <QuestionGenreScreen
          question={question}
          onAnswer={(userAnswer) => onUserAnswer(userAnswer, question, mistakes, maxMistakes)}
        />;

      case `artist`:
        return <ArtistQuestionScreen
          question={question}
          onAnswer={(userAnswer) => onUserAnswer(userAnswer, question, mistakes, maxMistakes)}
        />;
    }

    return null;
  }

  render() {
    const {
      questions,
      step,
    } = this.props;

    return <section className={`game ${Type.ARTIST}`}>
      <header className="game__header">
        <a className="game__back" href="#">
          <span className="visually-hidden">Сыграть ещё раз</span>
          <img className="game__logo" src="img/melody-logo-ginger.png" alt="Угадай мелодию"/>
        </a>

        <svg xmlns="http://www.w3.org/2000/svg" className="timer" viewBox="0 0 780 780">
          <circle className="timer__line" cx="390" cy="390" r="370" style={{filter: `url(#blur)`, transform: `rotate(-90deg) scaleY(-1)`, transformOrigin: `center`}}
          />
        </svg>

        <div className="timer__value" xmlns="http://www.w3.org/1999/xhtml">
          <span className="timer__mins">05</span>
          <span className="timer__dots">:</span>
          <span className="timer__secs">00</span>
        </div>

        <div className="game__mistakes">
          <div className="wrong"/>
          <div className="wrong"/>
          <div className="wrong"/>
        </div>
      </header>

      {this._getScreen(questions[step])}
    </section>;
  }
}

App.propTypes = {
  mistakes: PropTypes.number.isRequired,
  maxMistakes: PropTypes.number.isRequired,
  gameTime: PropTypes.number.isRequired,
  questions: PropTypes.array.isRequired,
  step: PropTypes.number.isRequired,
  onUserAnswer: PropTypes.func.isRequired,
  onWelcomeScreenClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  step: state.step,
  mistakes: state.mistakes,
});

const mapDispatchToProps = (dispatch) => ({
  onWelcomeScreenClick: () => dispatch(ActionCreator.incrementStep()),

  onUserAnswer: (userAnswer, question, mistakes, maxMistakes) => {
    dispatch(ActionCreator.incrementStep());
    dispatch(ActionCreator.incrementMistake(userAnswer, question, mistakes, maxMistakes));
  },
});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);
