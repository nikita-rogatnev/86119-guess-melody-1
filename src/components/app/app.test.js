import React from 'react';
import renderer from 'react-test-renderer';

import {App} from './app';

const mock = {
  questions: [
    {
      type: `genre`,
      genre: `rock`,
      answers: [
        {
          src: `test.mp3`,
          genre: `rock`,
        },
        {
          src: `test.mp3`,
          genre: `blues`,
        },
        {
          src: `test.mp3`,
          genre: `jazz`,
        },
        {
          src: `test.mp3`,
          genre: `rock`,
        },
      ],
    },
    {
      type: `artist`,
      song: {
        artist: `Jim Beam`,
        src: `path.mp3`,
      },
      answers: [
        {
          picture: `path.jpg`,
          artist: `John Snow`,
        },
        {
          picture: `path.jpg`,
          artist: `Jack Daniels`,
        },
        {
          picture: `path.jpg`,
          artist: `Jim Beam`,
        },
      ],
    }
  ],
};


it(`App correctly renders first screen`, () => {
  const {questions} = mock;
  const tree = renderer.create(<App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={-1}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  />).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`App correctly renders genre question screen`, () => {
  const {questions} = mock;
  const tree = renderer.create(<App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={1}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  />, {
    createNodeMock: () => {
      return {};
    }
  }).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`App correctly renders artist question screen`, () => {
  const {questions} = mock;
  const tree = renderer.create(<App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={2}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  />, {
    createNodeMock: () => {
      return {};
    }
  }).toJSON();

  expect(tree).toMatchSnapshot();
});
