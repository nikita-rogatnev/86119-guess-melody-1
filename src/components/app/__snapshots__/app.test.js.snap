// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders artist question screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
    </div>
  </header>
  <section
    className="welcome"
  >
    <div
      className="welcome__logo"
    >
      <img
        alt="Угадай мелодию"
        height="83"
        src="img/melody-logo.png"
        width="186"
      />
    </div>
    <button
      className="welcome__button"
      onClick={[MockFunction]}
    >
      <span
        className="visually-hidden"
      >
        Начать игру
      </span>
    </button>
    <h2
      className="welcome__rules-title"
    >
      Правила игры
    </h2>
    <p
      className="welcome__text"
    >
      Правила просты:
    </p>
    <ul
      className="welcome__rules-list"
    >
      <li>
        За 
        100
         минут нужно ответить на все вопросы.
      </li>
      <li>
        Можно допустить 
        Infinity
         ошибки.
      </li>
    </ul>
    <p
      className="welcome__text"
    >
      Удачи!
    </p>
  </section>
</section>
`;

exports[`App correctly renders first screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
    </div>
  </header>
  <section
    className="welcome"
  >
    <div
      className="welcome__logo"
    >
      <img
        alt="Угадай мелодию"
        height="83"
        src="img/melody-logo.png"
        width="186"
      />
    </div>
    <button
      className="welcome__button"
      onClick={[MockFunction]}
    >
      <span
        className="visually-hidden"
      >
        Начать игру
      </span>
    </button>
    <h2
      className="welcome__rules-title"
    >
      Правила игры
    </h2>
    <p
      className="welcome__text"
    >
      Правила просты:
    </p>
    <ul
      className="welcome__rules-list"
    >
      <li>
        За 
        100
         минут нужно ответить на все вопросы.
      </li>
      <li>
        Можно допустить 
        Infinity
         ошибки.
      </li>
    </ul>
    <p
      className="welcome__text"
    >
      Удачи!
    </p>
  </section>
</section>
`;

exports[`App correctly renders genre question screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
      <div
        className="wrong"
      />
    </div>
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="artist-0"
          name="answer"
          onClick={[Function]}
          type="radio"
          value="artist-0"
        />
        <label
          className="artist__name"
          htmlFor="artist-0"
        >
          <img
            alt="John Snow"
            className="artist__picture"
            src="path.jpg"
          />
          John Snow
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="artist-1"
          name="answer"
          onClick={[Function]}
          type="radio"
          value="artist-1"
        />
        <label
          className="artist__name"
          htmlFor="artist-1"
        >
          <img
            alt="Jack Daniels"
            className="artist__picture"
            src="path.jpg"
          />
          Jack Daniels
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id="artist-2"
          name="answer"
          onClick={[Function]}
          type="radio"
          value="artist-2"
        />
        <label
          className="artist__name"
          htmlFor="artist-2"
        >
          <img
            alt="Jim Beam"
            className="artist__picture"
            src="path.jpg"
          />
          Jim Beam
        </label>
      </div>
    </form>
  </section>
</section>
`;
